webhook.py

```diff
@@ -0,0 +1,171 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 SUSE LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import boto3
+import hashlib
+import hmac
+import json
+import os
+
+from base64 import b64decode
+
+
+def kms_decrypt(client, enc_value):
+    value = client.decrypt(
+        CiphertextBlob=b64decode(enc_value)
+    )['Plaintext']
+
+    return value
+
+
+kms_client = boto3.client('kms')
+
+EVENT_TYPES = ('ping', 'push',)
+SES_REGION = os.environ['SES_REGION']
+
+enc_secret = os.environ['REQUEST_SECRET']
+SECRET = kms_decrypt(kms_client, enc_secret)
+
+enc_email = os.environ['SOURCE_EMAIL']
+SOURCE_EMAIL = kms_decrypt(kms_client, enc_email).decode()
+
+
+def send_email(data):
+    repo = data['repository']
+    pusher = data['pusher']
+
+    empty_commit = '0000000000000000000000000000000000000000'
+    if data['before'] == empty_commit:
+        event = 'pushed a new branch'
+    elif data['after'] == empty_commit:
+        event = 'deleted a branch'
+    else:
+        event = 'pushed new commits'
+
+    subject = f'[{repo["full_name"]}] {data["ref"]}'
+    message = f'{pusher["name"]} {event} in {repo["full_name"]}\n\n' \
+              f'Branch: {data["ref"]}\n' \
+              f'Home:   {repo["html_url"]}\n' \
+              f'Pusher: {pusher["name"]} <{pusher["email"]}>\n' \
+              f'Diff:   {data["compare"]}\n\n'
+
+    for commit in data['commits']:
+        author = commit['author']
+        message += f'Commit: {commit["id"]}\n' \
+                   f'Message: {commit["message"]}\n' \
+                   f'URL: {commit["url"]}\n' \
+                   f'Author: {author["name"]} <{author["email"]}>\n' \
+                   f'Date: {commit["timestamp"]}\n\n'
+
+    try:
+        client = boto3.client(
+            'ses',
+            region_name=SES_REGION
+        )
+        client.send_email(
+            Source=SOURCE_EMAIL,
```

Comment:
This may cause a problem, with filtering depending on the how the receiving mail server is setup. For example a mail server can be configured such that it will reject messages from given source if that source does not match the IP of a known domain. SO if we set source-email to foo-team@suse.com via the command line argument it could be rejected since the e-mail is coming from an IP address that belongs to Amazon, since we know the IP of our outgoing mail server. It is safer to separate the source and target e-mails. As this is a github notification the sender should really be github.com

Reply:
Currently the email service sends emails with a source email == to the pushers email. This is not a good solution imho and triggers the same problem you mention. If you push commits the email shows up as sent from rjschwei@suse.com with the same reply-to.

In SES if you are not sandboxed you can send an email to any address without issue. However you can only send an email from a source if the email is validated as yours. So the source email has to be an email address we own where we can receive and act on a validation message.

Given the restrictions there's no way to use github.com as source unless you own a github.com email address.

And since the idea is the email goes to a mailing list I like the idea that it's all the same. If someone on the list wants to query the group about a push notification the email is already set to reply-to the same mailing list.

Reply:
https://docs.aws.amazon.com/ses/latest/DeveloperGuide/verify-email-addresses.html

Reply:
OK, fair enough, but this oddity belong sin the documentation somewhere then. We should not expect code readers, potential contributors to know about this "feature" of AWS SES. The info should either be in the README.md file or there should be a comment somewhere where this is being setup/used, probably in the code where the e-mail is queried.

Reply:
Added more info to README.